Sources/Rendering/WebGPU/UniformBuffer/index.js

```diff
@@ -0,0 +1,338 @@
+                entry.packed = true;
+                entry.offset = currOffset;
+                newEntries.push(entry);
+                currOffset += entry.sizeInBytes;
+                entry2.packed = true;
+                entry2.offset = currOffset;
+                newEntries.push(entry2);
+                currOffset += entry2.sizeInBytes;
+                entry3.packed = true;
+                entry3.offset = currOffset;
+                newEntries.push(entry3);
+                currOffset += entry3.sizeInBytes;
+                found = true;
+                break;
+              }
+            }
+          }
+        }
+      }
+    }
+
+    // Add anything remaining that is larger than 4 bytes and hope we get lucky.
+    // Likely if there is more than one item added here it will result
+    // in a failed UBO
+    for (let i = 0; i < model.bufferEntries.length; i++) {
+      const entry = model.bufferEntries[i];
+      if (!entry.packed && entry.sizeInBytes > 4) {
+        entry.packed = true;
+        entry.offset = currOffset;
+        newEntries.push(entry);
+        currOffset += entry.sizeInBytes;
+      }
+    }
+
+    // finally add remaining 4 byte items
+    for (let i = 0; i < model.bufferEntries.length; i++) {
+      const entry = model.bufferEntries[i];
+      if (!entry.packed) {
+        entry.packed = true;
+        entry.offset = currOffset;
+        newEntries.push(entry);
+        currOffset += entry.sizeInBytes;
+      }
+    }
+
+    // update entries and entryNames
+    model.bufferEntries = newEntries;
+    model.bufferEntryNames.clear();
+    for (let i = 0; i < model.bufferEntries.length; i++) {
+      model.bufferEntryNames.set(model.bufferEntries[i].name, i);
+    }
+    model.sizeInBytes = currOffset;
+    model.sortDirty = false;
+  };
+
+  publicAPI.sendIfNeeded = (device, layout) => {
+    if (!model.UBO) {
+      const req = {
+        address: model.Float32Array,
+        time: 0,
+        usage: BufferUsage.UniformArray,
+      };
+      model.UBO = device.getBufferManager().getBuffer(req);
+      model.bindGroup = device.getHandle().createBindGroup({
+        layout,
+        entries: [
+          {
+            binding: model.binding,
+            resource: {
+              buffer: model.UBO.getHandle(),
+            },
+          },
+        ],
+      });
+      model.sendTime.modified();
+      model.sendDirty = false;
+    }
+
+    // send data down if needed
+    if (model.sendDirty) {
+      device
+        .getHandle()
+        .queue.writeBuffer(
+          model.UBO.getHandle(),
+          0,
+          model.arrayBuffer,
+          0,
+          model.sizeInBytes
+        );
+      model.sendTime.modified();
+      model.sendDirty = false;
+    }
+  };
+
+  publicAPI.createView = (type) => {
+    if (type in model === false) {
+      if (!model.arrayBuffer) {
+        model.arrayBuffer = new ArrayBuffer(model.sizeInBytes);
+      }
+      model[type] = new window[type](model.arrayBuffer);
```

Comment:
You should not use `window[type]` but `TYPED_ARRAYS[type]` instead from `import { TYPED_ARRAYS } from 'vtk.js/Sources/macro';`.

The reason being that `window` does not exist outside of a browser. That way we can support some usage of vtk.js inside node/deno and else...

Reply:
OK I just copied it form DataArray. Should I change DataArray as well?

Reply:
yes if you can. The `TYPED_ARRAYS` was added later when some folks used vtk.js on node and it was not working...

Reply:
OK that was a large change. There were A LOT of window[dataType] uses in the code (which could have been changed when TYPED_ARRAYS was added). Plus some folks were passing Array as a data type which window[Array] worked but TYPED_ARRAYS[Array] does not. We really do not want non-native arrays in DataArray so I made some changes to handle that case as well.